twine/settings.py

```diff
@@ -0,0 +1,280 @@
+
+
+class Settings(object):
+    """Object that manages the configuration for Twine.
+
+    This object can only be instantiated with keyword arguments.
+
+    For example,
+
+    .. code-block:: python
+
+        Settings(True, username='fakeusername')
+
+    Will raise a :class:`TypeError`. Instead, you would want
+
+    .. code-block:: python
+
+        Settings(sign=True, username='fakeusername')
+    """
+
+    @utils.no_positional(allow_self=True)
+    def __init__(self,
+                 sign=False, sign_with='gpg', identity=None,
+                 username=None, password=None,
+                 comment=None,
+                 config_file='~/.pypirc', skip_existing=False,
+                 cacert=None, client_cert=None,
+                 repository_name='pypi', repository_url=None,
+                 **ignored_kwargs
+                 ):
+        """Initialize our settings instance.
+
+        :param bool sign:
+            Configure whether the package file should be signed.
+
+            This defaults to ``False``.
+        :param str sign_with:
+            The name of the executable used to sign the package with.
+
+            This defaults to ``gpg``.
+        :param str identity:
+            The GPG identity that should be used to sign the package file.
+        :param str username:
+            The username used to authenticate to the repository (package
+            index).
+        :param str password:
+            The password used to authenticate to the repository (package
+            index).
+        :param str comment:
+            The comment to include with each distribution file.
+        :param str config_file:
+            The path to the configuration file to use.
+
+            This defaults to ``~/.pypirc``.
+        :param bool skip_existing:
+            Specify whether twine should continue uploading files if one
+            of them already exists. This primarily supports PyPI. Other
+            package indexes may not be supported.
+
+            This defaults to ``False``.
+        :param str cacert:
+            The path to the bundle of certificates used to verify the TLS
+            connection to the package index.
+        :param str client_cert:
+            The path to the client certificate used to perform authentication
+            to the index.
+
+            This must be a single file that contains both the private key and
+            the PEM-encoded certificate.
+        :param str repository_name:
+            The name of the repository (package index) to interact with. This
+            should correspond to a section in the config file.
+        :param str repository_url:
+            The URL of the repository (package index) to interact with. This
+            will override the settings inferred from ``repository_name``.
+        """
+        self.config_file = config_file
+        self.comment = comment
+        self.skip_existing = skip_existing
+        self._handle_repository_options(
```

Comment:
I think these methods will be the most controversial part of this.

On the one hand, I can imagine having a separate object, e.g., `SettingsBuilder`, that does this work and then makes `Settings` into a mostly data class that stores the validated/correct configuration and does the other nice things (like checking repository URLs and creating a new repository). I went with the simpler/thinner implementation to start with given the lack of feedback on the API design document, though.